# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [2]:
import pandas as pd
import numpy as np
import sqlite3

/tmp/ipykernel_6096/4081858790.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [3]:
db = sqlite3.connect('data/go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [4]:
tablenames = pd.read_sql_query("SELECT name FROM sqlite_master WHERE type='table' AND name NOT IN ('sqlite_sequence');", db)
tables: dict[str, pd.DataFrame]= {}
for table in tablenames['name']:
    tables[table] = pd.read_sql_query(f"SELECT * FROM {table};", db)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [5]:
products = tables['product']
products["PRODUCTION_COST"] = products["PRODUCTION_COST"].astype(float)
filter = (products["PRODUCTION_COST"] > 50) & (products["PRODUCTION_COST"] < 100)
filtered = products.loc[(filter), :]
filtered

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE  PRODUCTION_COST MARGIN  \
13            110        10-12-2019                20            60.00     .5   
14            111        15-12-2019                20            81.80    .55   
17            114        15-12-2019                21            80.00     .6   
24             17          5-3-2013                 3            60.00    .28   
25             18         15-2-2011                 3            86.00    .33   
28             20          5-3-2013                 3            78.55    .28   
31             23          5-3-2013                 3            65.33    .33   
32             24          5-3-2013                 4            62.50    .17   
56             46         10-1-2021                 7            52.86    .28   
58             48         10-1-2021                 7            55.00     .5   
69             58         10-1-2021                 9            57.13    .28   
71              6          5-3-2013                 1            85.11    .28   
81             69         15-2-2011                11            73.33    .33   
83             70          5-3-2013                11            80.00    .33   
84             71          5-3-2013                11            60.00    .33   
85             72        26-10-2016                11            72.50     .5   
88             75        26-10-2016                12            80.00    .28   
91             78         15-2-2011                13            79.19     .2   
92             79         15-2-2011                13            92.58    .25   
93              8          5-3-2013                 1            75.00    .41   
94             80          5-3-2013                13            94.12     .4   
97             83          5-3-2013                14            56.25    .37   
98             84          5-3-2013                14            78.55    .28   

      PRODUCT_IMAGE LANGUAGE                     PRODUCT_NAME  \
13  P110GE5PT20.jpg       EN                Blue Steel Putter   
14  P111GE5PT20.jpg       EN            Blue Steel Max Putter   
17  P114GE5GA21.jpg       EN              Course Pro Golf Bag   
24    P17CE1SB3.jpg       EN                  Hibernator Lite   
25    P17CE1SB3.jpg       EN                       Hibernator   
28    P20CE1SB3.jpg       EN  Hibernator Self - Inflating Mat   
31    P23CE1SB3.jpg       EN              Hibernator Camp Cot   
32    P24CE1PK4.jpg       EN     Canyon Mule Climber Backpack   
56    P46ME2SY7.jpg       EN          Granite Climbing Helmet   
58    P47ME2SY7.jpg       EN            Husky Harness Extreme   
69    P58ME2TL9.jpg       EN                   Granite Hammer   
71    P06CE1CG1.jpg       EN        TrailChef Deluxe Cook Set   
81   P69PA3EW11.jpg       EN                        Polar Ice   
83   P70PA3EW11.jpg       EN                     Polar Sports   
84   P71PA3EW11.jpg       EN                       Polar Wave   
85   P72PA3EW11.jpg       EN                    Polar Extreme   
88   P75PA3KV12.jpg       EN                     Edge Extreme   
91   P78PA3BN13.jpg       EN                        Seeker 35   
92   P78PA3BN13.jpg       EN                        Seeker 50   
93    P08CE1CG1.jpg       EN           TrailChef Double Flame   
94   P80PA3BN13.jpg       EN                   Seeker Extreme   
97   P83PA3NV14.jpg       EN                   Glacier Deluxe   
98   P84PA3NV14.jpg       EN                      Glacier GPS   

                                          DESCRIPTION TRIAL888  
13  This premium putter has copper mixed with ligh...        T  
14  Putter head is composed from a single piece of...        T  
17  High quality golf bag that includes carrying s...        T  
24  The Hibernator Lite is the perfect summer slee...        T  
25  The Hibernator is a three-season sleeping bag....        T  
28  Mattress is made of extra-thick foam. Mesh stu...        T  
31  Aluminum frame camp cot that is lightw

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [6]:
products["MARGIN"] = products["MARGIN"].astype(float)
filter = (products["MARGIN"] < 0.2) | (products["MARGIN"] > 0.6)
filtered = products.loc[(filter), :]
filtered

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE  PRODUCTION_COST  \
29              21         15-2-2011                 3            29.17   
32              24          5-3-2013                 4            62.50   
100             86         15-2-2011                15             2.00   
101             87         15-2-2011                15             2.00   
102             88         15-2-2011                15             2.00   
103             89         15-2-2011                15             2.50   
105             90         15-2-2011                15             2.50   

     MARGIN   PRODUCT_IMAGE LANGUAGE                  PRODUCT_NAME  \
29     0.17   P21CE1SB3.jpg       EN                Hibernator Pad   
32     0.17   P24CE1PK4.jpg       EN  Canyon Mule Climber Backpack   
100    0.67  P86OP4IR15.jpg       EN             BugShield Natural   
101    0.67  P86OP4IR15.jpg       EN               BugShield Spray   
102    0.70  P86OP4IR15.jpg       EN         BugShield Lotion Lite   
103    0.63  P86OP4IR15.jpg       EN              BugShield Lotion   
105    0.63  P86OP4IR15.jpg       EN             BugShield Extreme   

                                           DESCRIPTION TRIAL888  
29   Warms up quickly for snow camping comfort. The...        T  
32   This pack is perfect for day trips and short h...        T  
100  Contains no chemicals. Repels mosquitoes, blac...        T  
101  Repels black flies, ticks, mosquitoes, chigger...        T  
102  Repels black flies, ticks, mosquitoes, chigger...        T  
103  Repels black flies, ticks, mosquitoes, chigger...        T  
105  Highest amount of the active ingredient, 38%, ...        T

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [7]:
countries = tables['country']
filter  = countries["CURRENCY_NAME"] == "francs"
filtered = countries.loc[(filter), :]
filtered

COUNTRY_CODE      COUNTRY LANGUAGE CURRENCY_NAME TRIAL879
0             1       France       EN        francs        T
7            17      Belgium       EN        francs        T
19            8  Switzerland       EN        francs        T

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [8]:
filtered = products.loc[(products["MARGIN"] > 0.5), :].drop_duplicates("INTRODUCTION_DATE")
filtered

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE  PRODUCTION_COST  \
4              102        10-12-2019                18           380.95   
6              104        18-12-2019                18           333.32   
7              105        27-12-2019                19           555.90   
9              107         13-1-2020                19           509.07   
14             111        15-12-2019                20            81.80   
79              67        26-10-2016                10           116.19   
100             86         15-2-2011                15             2.00   

     MARGIN    PRODUCT_IMAGE LANGUAGE                       PRODUCT_NAME  \
4      0.51  P102GE5IR18.jpg       EN           Hailstorm Titanium Irons   
6      0.51  P102GE5IR18.jpg       EN      Lady Hailstorm Titanium Irons   
7      0.55  P105GE5WD19.jpg       EN       Hailstorm Titanium Woods Set   
9      0.55  P105GE5WD19.jpg       EN  Lady Hailstorm Titanium Woods Set   
14     0.55  P111GE5PT20.jpg       EN              Blue Steel Max Putter   
79     0.51   P67PA3WT10.jpg       EN               Mountain Man Extreme   
100    0.67   P86OP4IR15.jpg       EN                  BugShield Natural   

                                           DESCRIPTION TRIAL888  
4    Made entirely of pure titanium. The ultimate i...        T  
6    Made entirely of pure titanium. A wider sole a...        T  
7    Crafted of 6-4 titanium which is 40% lighter a...        T  
9    Crafted of 6-4 titanium which is 40% lighter a...        T  
14   Putter head is composed from a single piece of...        T  
79   Solid titanium case, crown, bezel.  Combinatio...        T  
100  Contains no chemicals. Repels mosquitoes, blac...        T

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [9]:
sales_branches = tables['sales_branch']
filter = sales_branches["ADDRESS2"].notna() & sales_branches["REGION"].notna()
filtered = sales_branches.loc[(filter), :]
filtered

SALES_BRANCH_CODE                          ADDRESS1           ADDRESS2  \
10                24  Prol. Paseo de la Reforma No. 51  Lomas de Santa Fe   
11                25               202-2-3 Hyakunincho        Shinjuku-ku   
12                26                     543-225 Asahi           Asahi-ku   
13                28                  2315 Queen's Ave            Level 2   
15                30             Avenida Paulista, 333   CJ 231 2o. Andar   
21                36                    3 Albany Court      Market Street   
27                 9               Singelgravenplein 4      4e verdieping   

                CITY         REGION POSTAL_ZONE COUNTRY_CODE TRIAL888  
10  Distrito Federal         México       01219           20        T  
11             Tokyo          Tokyo    681-3187           11        T  
12        Osaka City          Osaka    616-5698           11        T  
13         Melbourne            VIC        2088           16        T  
15         São Paulo             SP   01403-090           21        T  
21        Birmingham  West Midlands     B43 7NP            9        T  
27         Amsterdam  Noord-Holland     1233 BW            7        T

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [10]:
filter = (countries["CURRENCY_NAME"] == "dollars") | (countries["CURRENCY_NAME"] == "new dollar")
filtered = countries.loc[(filter), "COUNTRY"]
filtered

3            Taiwan
6         Australia
14    United States
15           Canada
Name: COUNTRY, dtype: object

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [11]:
retailer_sites = tables['retailer_site']
filter = (retailer_sites["POSTAL_ZONE"].str[0] == "D") & (retailer_sites["ADDRESS2"].notna())
filtered = retailer_sites.loc[(filter), ['ADDRESS1', 'ADDRESS2', 'POSTAL_ZONE']]
filtered

ADDRESS1       ADDRESS2 POSTAL_ZONE
205  Röntgenstraße 90  3. Tür rechts     D-60314
351   Grubesallee 141       4. Stock     D-22335

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [12]:
returned_items = tables['returned_item']
total = returned_items["RETURN_CODE"].count()
total

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [13]:
copy = sales_branches.drop_duplicates("REGION")
total = copy["REGION"].count()
total

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [14]:
results = products["MARGIN"].agg(["min", "max", "mean"]).round(2)
results

min     0.17
max     0.70
mean    0.40
Name: MARGIN, dtype: float64

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [15]:
no_second_address = retailer_sites["ADDRESS2"].isna()
total = no_second_address.count()
total

391

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [16]:
order_details = tables['order_details']
order_details["UNIT_SALE_PRICE"] = order_details["UNIT_SALE_PRICE"].astype(float)
order_details["UNIT_PRICE"] = order_details["UNIT_PRICE"].astype(float)

filter = order_details["UNIT_SALE_PRICE"] < order_details["UNIT_PRICE"]
filtered = order_details.loc[(filter), :]

for index, row in filtered.iterrows():
    filtered.at[index, "DIFFERENCE"] = row["UNIT_PRICE"] - row["UNIT_SALE_PRICE"]

average = filtered["DIFFERENCE"].agg(np.mean)
average


/tmp/ipykernel_6096/2782065147.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered.at[index, "DIFFERENCE"] = row["UNIT_PRICE"] - row["UNIT_SALE_PRICE"]


/tmp/ipykernel_6096/2782065147.py:11: FutureWarning: The provided callable <function mean at 0x7fbe54598ca0> is currently using Series.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  average = filtered["DIFFERENCE"].agg(np.mean)


24.236185630800964

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [17]:
sales_staff = tables['sales_staff']
per_function = sales_staff.groupby("POSITION_EN", as_index=False).count()
newtable = per_function.loc[:, ["POSITION_EN", "SALES_STAFF_CODE"]]
newtable.rename(columns={"SALES_STAFF_CODE": "TOTAL"}, inplace=True)
newtable

POSITION_EN  TOTAL
0                Branch Manager     19
1              District Manager      4
2               General Manager      1
3  Level 1 Sales Representative     13
4  Level 2 Sales Representative     39
5  Level 3 Sales Representative     24
6              Regional Manager      2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [18]:
count_per_phonenumber = (sales_staff.groupby("WORK_PHONE", as_index=False)
                       .agg(COUNT=('SALES_STAFF_CODE', 'size'))
                       .query('COUNT > 4')[["WORK_PHONE", "COUNT"]])
count_per_phonenumber

WORK_PHONE  COUNT
4     +(44) 121 3505267      5
10  +(61) 03 2982  4242      5
11  +31 (0)20 692 93 94      5
13    +33 1 68 94 52 20      5
18     1 (206) 292-0012      5
19     1 (305) 557-4810      6
20     1 (310) 281-5722      5
21     1 (403) 232-5986      5
22     1 (416) 493-5595      5
23     1 (617) 268-6754      6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [19]:
is_netherlands = countries["COUNTRY"] == "Netherlands"
filtered = countries.loc[(is_netherlands), :]
merged = pd.merge(retailer_sites, filtered, "inner", "COUNTRY_CODE")[["ADDRESS1", "CITY"]]
merged

ADDRESS1          CITY
0                    Beets 36         Beets
1                Perestraat 1      Deventer
2   Laan van Meerdervoort 966      Den Haag
3              Kalfjeslaan 47    Amstelveen
4                Startbaan 11    Amstelveen
5       Noorderspoorsingel 28     Groningen
6               Hogehilweg 13     Amsterdam
7              Spoorstraat 43    Varsseveld
8              Westhavenweg 8     Amsterdam
9               Westzeedijk 4        Muiden
10            Gooiergracht 91         Laren
11            Hoofdstraat 150   Pieterburen
12                Olde Diek 2         Laren
13     Amsterdamsestraatweg 8      Maarssen
14        Krayenhoffstraat 11     Amsterdam
15         Overschiestraat 19     Amsterdam
16              Havenbuurt 12        Marken
17               Duivendaal 1    Wageningen
18              Voorstraat 37  Nieuweschans
19         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [20]:
product_types = tables["product_type"]
has_eyewear = product_types.loc[(product_types["PRODUCT_TYPE_EN"] == "Eyewear"), :]
merged = pd.merge(products, has_eyewear, "inner", "PRODUCT_TYPE_CODE")[["PRODUCT_NAME"]]
merged

PRODUCT_NAME
0      Polar Sun
1      Polar Ice
2   Polar Sports
3     Polar Wave
4  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [21]:
order_headers = tables["order_header"]

branch_managers = sales_staff.loc[(sales_staff["POSITION_EN"] == "Branch Manager"), :]
retailers_with_orders = pd.merge(retailer_sites, order_headers, "inner", "RETAILER_SITE_CODE")
retailers_with_orders = retailers_with_orders.drop_duplicates(["ADDRESS1", "SALES_STAFF_CODE"])

retailers_orders_with_sales_staff = pd.merge(retailers_with_orders, branch_managers, "inner", "SALES_STAFF_CODE")
retailers_orders_with_sales_staff[["ADDRESS1", "FIRST_NAME", "LAST_NAME"]]

ADDRESS1 FIRST_NAME LAST_NAME
0  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [22]:
is_manager = sales_staff["POSITION_EN"].str.contains("Manager")
managers = sales_staff.loc[(is_manager), ["POSITION_EN", "SALES_STAFF_CODE"]]

selling_managers = pd.merge(managers, order_headers, "left", "SALES_STAFF_CODE")[["POSITION_EN", "ORDER_DATE"]]
selling_managers.drop_duplicates(["POSITION_EN", "ORDER_DATE"])

POSITION_EN  ORDER_DATE
0     Branch Manager         NaN
1    General Manager         NaN
4   Regional Manager         NaN
7   District Manager         NaN
25    Branch Manager  2022-09-09
26    Branch Manager  2022-10-31
27    Branch Manager  2022-12-13
28    Branch Manager  2022-03-13
29    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [23]:
mass_productcodes = order_details.loc[(order_details["QUANTITY"] > 750), "PRODUCT_NUMBER"].drop_duplicates()
mass_products = pd.merge(products, mass_productcodes, "inner", "PRODUCT_NUMBER")[["PRODUCT_NAME", "PRODUCT_TYPE_CODE"]]
mass_products_with_typename = pd.merge(mass_products, product_types, "inner", "PRODUCT_TYPE_CODE")[["PRODUCT_NAME", "PRODUCT_TYPE_EN"]]
mass_products_with_typename

PRODUCT_NAME    PRODUCT_TYPE_EN
0               Star Peg              Tents
1              Firefly 2           Lanterns
2        BugShield Spray  Insect Repellents
3  BugShield Lotion Lite  Insect Repellents
4       BugShield Lotion  Insect Repellents
5      BugShield Extreme  Insect Repellents
6         Sun Shelter 15          Sunscreen
7         Sun Shelter 30          Sunscreen
8             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [24]:
filter = ((order_details["UNIT_PRICE"] - order_details["UNIT_SALE_PRICE"]) / order_details["UNIT_PRICE"]) > 0.4
filtered = order_details.loc[(filter), ["PRODUCT_NUMBER"]].drop_duplicates()
result = pd.merge(filtered, products, "inner", "PRODUCT_NUMBER")[["PRODUCT_NAME"]]
result

PRODUCT_NAME
0         BugShield Lotion Lite
1             BugShield Extreme
2             BugShield Natural
3                Sun Shelter 30
4              BugShield Lotion
5               BugShield Spray
6  Hailstorm Titanium Woods Set
7                Sun Shelter 15

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [25]:
returned_items["RETURN_QUANTITY"] = returned_items["RETURN_QUANTITY"].astype(float)

return_with_orderdetail = pd.merge(returned_items, order_details, "inner", "ORDER_DETAIL_CODE")
filter = (return_with_orderdetail["RETURN_QUANTITY"] / return_with_orderdetail["QUANTITY"]) > 0.9
filtered = return_with_orderdetail.loc[(filter), ["RETURN_REASON_CODE"]].drop_duplicates()

return_reasons = tables["return_reason"]
reasons = pd.merge(filtered, return_reasons, "inner", "RETURN_REASON_CODE")[["RETURN_DESCRIPTION_EN"]]
reasons

RETURN_DESCRIPTION_EN
0  Unsatisfactory product
1   Wrong product shipped
2   Wrong product ordered

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [26]:
products_with_type = pd.merge(products, product_types, "inner", "PRODUCT_TYPE_CODE")
products_per_type  = products_with_type.groupby("PRODUCT_TYPE_EN", as_index=False).count()[["PRODUCT_TYPE_EN", "PRODUCT_NUMBER"]]
products_per_type.rename(columns={"PRODUCT_NUMBER": "TOTAL"})

PRODUCT_TYPE_EN  TOTAL
0             Binoculars      4
1   Climbing Accessories      7
2           Cooking Gear     10
3                Eyewear      5
4              First Aid      5
5       Golf Accessories      4
6      Insect Repellents      5
7                  Irons      4
8                 Knives      5
9               Lanterns     12
10            Navigation      4
11                 Packs      6
12               Putters      3
13                  Rope      4
14                Safety      4
15         Sleeping Bags      7
16             Sunscreen      5
17                 Tents      6
18                 Tools      6
19               Watches      5
20                 Woods      4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [27]:
sites_with_countries = pd.merge(retailer_sites, countries, "inner", "COUNTRY_CODE")[["COUNTRY", "RETAILER_SITE_CODE"]]
sites_per_country = sites_with_countries.groupby("COUNTRY", as_index=False).count()[["COUNTRY", "RETAILER_SITE_CODE"]]
sites_per_country.rename(columns={"RETAILER_SITE_CODE": "TOTAL"})

COUNTRY  TOTAL
0        Australia     10
1          Austria     10
2          Belgium     10
3           Brazil      1
4           Canada     31
5            China     10
6          Denmark      6
7          Finland      5
8           France     29
9          Germany     32
10           Italy     12
11           Japan     25
12           Korea     10
13          Mexico     10
14     Netherlands     20
15           Spain      9
16          Sweden     12
17     Switzerland     13
18          Taiwan     10
19  United Kingdom     25
20   United States    101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [28]:
cooking_gear_products = products_with_type.loc[(products_with_type["PRODUCT_TYPE_EN"] == "Cooking Gear"), :][["PRODUCT_NAME", "PRODUCT_NUMBER"]]

total_sale_quantity_per_product = (
    pd.merge(cooking_gear_products, order_details, "inner", "PRODUCT_NUMBER")
    .groupby("PRODUCT_NAME", as_index=False)
    .agg(TOTAL_SALE_QUANTITY=("QUANTITY", "sum"))
)

total_sale_quantity_per_product



PRODUCT_NAME  TOTAL_SALE_QUANTITY
0          TrailChef Canteen                31712
1         TrailChef Cook Set                12700
2              TrailChef Cup                31834
3  TrailChef Deluxe Cook Set                 3744
4     TrailChef Double Flame                14510
5           TrailChef Kettle                 9900
6      TrailChef Kitchen Kit                25964
7     TrailChef Single Flame                10614
8         TrailChef Utensils                15928
9        TrailChef Water Bag                41770

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [29]:
sales_targets = tables["SALES_TARGETData"]

amsterdam_branches = sales_branches.loc[(sales_branches["CITY"] == "Amsterdam"), "SALES_BRANCH_CODE"]
amsterdam_staff    = sales_staff.loc[(sales_staff["SALES_BRANCH_CODE"].isin(amsterdam_branches)), :]

targets_2022       = sales_targets.loc[(sales_targets["SALES_YEAR"] == "2022"), :]
staff_with_targets = pd.merge(amsterdam_staff, targets_2022, "inner", "SALES_STAFF_CODE")

staff_target_groups = (
    staff_with_targets.groupby(["SALES_STAFF_CODE", "FIRST_NAME", "LAST_NAME"], as_index=False)
    .agg(LOWEST_TARGET=("SALES_TARGET", "min"), HIGHEST_TARGET=("SALES_TARGET", "max"))
)

staff_target_groups[['FIRST_NAME', 'LAST_NAME', 'LOWEST_TARGET', 'HIGHEST_TARGET']]

FIRST_NAME         LAST_NAME LOWEST_TARGET HIGHEST_TARGET
0    Belinda  Jansen-Velasquez          1005            963
1      Ellen           Shapiro          1004            986
2        Jan         Haverkamp          1048            993
3      Maria          Laponder          1000           9956

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [30]:
branches_with_country = pd.merge(sales_branches, countries, "inner", "COUNTRY_CODE")[["SALES_BRANCH_CODE", "CITY", "COUNTRY", "COUNTRY_CODE"]]
branches_with_country = branches_with_country.rename(columns={"CITY": "VERKOPER"})

branches_with_sellers = pd.merge(branches_with_country, retailer_sites, "inner", "COUNTRY_CODE")
branches_sellers_per_country = branches_with_sellers.groupby(["COUNTRY", "VERKOPER"], as_index=False).agg(KLANTEN=("RETAILER_CODE", "count"))

branches_sellers_per_country

COUNTRY          VERKOPER  KLANTEN
0        Australia         Melbourne       10
1          Austria              Wien       10
2          Belgium          Heverlee       10
3           Brazil         São Paulo        1
4           Canada           Calgary       31
5           Canada           Toronto       31
6            China          Shanghai       10
7          Finland            Kuopio        5
8           France              Lyon       29
9           France             Paris       29
10         Germany           Hamburg       32
11         Germany           München       32
12           Italy            Milano       12
13           Japan        Osaka City       25
14           Japan             Tokyo       25
15           Korea             Seoul       10
16          Mexico  Distrito Federal       10
17     Netherlands         Amsterdam       20
18           Spain            Bilbao        9
19          Sweden             Kista       12
20     Switzerland            Zürich       13
21          Taiwan         San Chung       10
22  United Kingdom        Birmingham       25
23  United Kingdom            London       25
24   United States            Boston      101
25   United States       Los Angeles      101
26   United States             Miami      101
27   United States           Seattle      101

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [31]:
sellers = order_headers.copy().drop_duplicates("SALES_STAFF_CODE").reset_index(drop=True)
filter = sales_staff["SALES_STAFF_CODE"].isin(sellers["SALES_STAFF_CODE"])
not_sellers = sales_staff.loc[(~filter), :]
not_sellers


SALES_STAFF_CODE FIRST_NAME      LAST_NAME       POSITION_EN  \
7               107     Walter         Taylor    Branch Manager   
8               108   Penelope        Tamrine   General Manager   
11              111      Polly         Linton    Branch Manager   
19              119      Gracy        Gellens    Branch Manager   
26               13       Else         Mörike  Regional Manager   
27               14      Frank          Fuchs    Branch Manager   
30               17      Fritz         Hirsch    Branch Manager   
31               18       Jörg          Kunze  District Manager   
33               20      Maria      Iacobucci    Branch Manager   
38               25       Kick        Kalkman    Branch Manager   
43               30      Karin      Bergström    Branch Manager   
44                4      Denis           Pagé    Branch Manager   
47               45      Sally          White    Branch Manager   
49               47      Frank        Bretton  District Manager   
50               48    Georges  Saint-Germain    Branch Manager   
56               53       Alex      Rodriguez    Branch Manager   
60               57   Samantha          Floyd    Branch Manager   
65               61      Julie          Olsen    Branch Manager   
69               65      Corey         Wright    Branch Manager   
76               71    Estelle        Johnson  District Manager   
77               72       Greg        Belding  Regional Manager   
80               75      Julie        Béchard    Branch Manager   
84               79      Lucía          Reyna    Branch Manager   
89               87     Kazumi        Uragome  District Manager   
91               89      Alice         Walter    Branch Manager   

             WORK_PHONE EXTENSION                  FAX  \
7     +(44) 181 2033144      3547    +(44) 181 5030282   
8     +(44) 181 2033144      3542    +(44) 181 5030282   
11    +(44) 121 3505267      1216    +(44) 121 3583843   
19      +32 16 20.73.21      1349      +32 16 20.73.32   
26    +(49) 40 663 1990      1865    +(49) 40 663 4571   
27    +(49) 40 663 1990      1847    +(49) 40 663 4571   
30    +(49) 89 882 3456       225    +(49) 89 882 3455   
31    +(49) 89 882 3456       234    +(49) 89 882 3455   
33   +(39) 02 41 05 400      None   +(39) 02 41 20 549   
38  +31 (0)20 692 93 94       197  +31 (0)20 692 93 06   
43     +(46) 08 753 116       215     +(46) 08 752 592   
44    +33 1 68 94 52 20       325    +33 1 68 94 56 60   
47     1 (403) 232-5986       317     1 (403) 232-5831   
49     1 (403) 232-5986       301     1 (403) 232-5831   
50     1 (416) 493-5595      1245     1 (416) 496-9667   
56     1 (617) 268-6754      1334     1 (617) 268-3464   
60     1 (206) 292-0012       331     1 (206) 292-3312   
65     1 (310) 281-5722      2045     1 (310) 281-9087   
69     1 (305) 557-4810      3490     1 (305) 557-7650   
76     1 (617) 268-6754      1399     1 (617) 268-3464   
77     1 (206) 292-0012       373     1 (206) 292-3312   
80    +33 4 52 26 33 88      1315    +33 4 66 35 98 95   
84      +(525) 549 6090       452       +(52) 549 6088   
89        81-6-616-5656      2156        81-6-616-5657   
91  +(61) 03 2982  4242      1845  +(61) 03 2982  4243   

                        EMAIL  DATE_HIRED SALES_BRANCH_CODE TRIAL888  
7         WTaylor@grtd123.com    9-6-2011                35        T  
8        PTamrine@grtd123.com  18-12-2014                35        T  
11        PLinton@grtd123.com   18-6-2012                36        T  
19       GGellens@grtd123.com   10-9-2015                38        T  
26        EMorike@grtd123.com    2-6-2015                13        T  
27         FFuchs@grtd123.com   12-1-2013                13        T  
30        FHirsch@grtd123.com   12-9-2012                14        T  
31         JKunze@grtd123.com    7-1-2016                14        T  
33     MIacobucci@grtd123.com   18-8-2012                 7        T  
38       KKalkman@grtd123.com    6-9-2011

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [48]:
average_margin = products["MARGIN"].mean()
lower_than_average = products.loc[(products["MARGIN"] < average_margin), :]
count_and_average = lower_than_average.agg(COUNT=("MARGIN", "count"), AVERAGE=("MARGIN", "mean"))
count_and_average

MARGIN
COUNT    59.000000
AVERAGE   0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [54]:
unit_sale_price_above_500 = order_details.loc[(order_details["UNIT_SALE_PRICE"] > 500), :]
unit_sale_price_above_500 = unit_sale_price_above_500.drop_duplicates("PRODUCT_NUMBER")
unit_sale_price_above_500
products_with_high_salesprices = pd.merge(products, unit_sale_price_above_500, "inner", "PRODUCT_NUMBER")[["PRODUCT_NAME"]]
products_with_high_salesprices

PRODUCT_NAME
0               Hailstorm Steel Irons
1            Hailstorm Titanium Irons
2          Lady Hailstorm Steel Irons
3       Lady Hailstorm Titanium Irons
4        Hailstorm Titanium Woods Set
5           Hailstorm Steel Woods Set
6   Lady Hailstorm Titanium Woods Set
7      Lady Hailstorm Steel Woods Set
8                           Star Dome
9                        Star Gazer 2
10                       Star Gazer 3
11                       Star Gazer 6
12                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [58]:
sales_staff_copy = sales_staff.copy()
for index, row in sales_staff_copy.iterrows():
    is_manager = "Manager" in row["POSITION_EN"]
    sales_staff_copy.at[index, "IS_MANAGER"] = "Ja" if is_manager else "Nee"

sales_staff_copy[["LAST_NAME", "IS_MANAGER"]]

LAST_NAME IS_MANAGER
0    Savolainen        Nee
1           Kim        Nee
2          Choi        Nee
3           Kao        Nee
4          Fang        Nee
..          ...        ...
97     Iglesias        Nee
98        Lopes         Ja
99       Barros        Nee
100   Guimarães        Nee
101    Lindholm        Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [59]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [34]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [62]:
sales_staff_copy = sales_staff.copy()
for index, row in sales_staff_copy.iterrows():
    date_hired = datetime.strptime(row["DATE_HIRED"], '%d-%m-%Y')
    time_employed = datetime.today() - date_hired
    years_employed = time_employed.days / 365
    sales_staff_copy.at[index, "DIENST_TIJD"] = "Lang in dienst" if years_employed >= 12 else "Kort in dienst"

sales_staff_copy[["LAST_NAME", "DIENST_TIJD"]]



LAST_NAME     DIENST_TIJD
0    Savolainen  Kort in dienst
1           Kim  Kort in dienst
2          Choi  Kort in dienst
3           Kao  Kort in dienst
4          Fang  Kort in dienst
..          ...             ...
97     Iglesias  Kort in dienst
98        Lopes  Kort in dienst
99       Barros  Kort in dienst
100   Guimarães  Kort in dienst
101    Lindholm  Kort in dienst

[102 rows x 2 columns]